<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Vincent Roh" />

<meta name="date" content="2017-10-20" />

<title>BC32BarSeq Vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">BC32BarSeq Vignette</h1>
<h4 class="author"><em>Vincent Roh</em></h4>
<h4 class="date"><em>20 Oct 2017</em></h4>



<p>This is the instruction vignette for the package BC32BarSeq. It is designed to help analyze BarSeq sequencing runs, specifically with the BC32 system. The set of functions provided in this package depends on the following R packages, so make sure they are installed:</p>
<ul>
<li>ShortRead</li>
<li>Biostrings</li>
<li>dplyr</li>
<li>stringdist</li>
<li>ggplot2</li>
<li>gridExtra</li>
<li>msa</li>
<li>shiny</li>
<li>reshape</li>
</ul>
<p>Load the BC32BarSeq package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BC32BarSeq)</code></pre></div>
<p>Load all other packages at once with the function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">initBC</span>()</code></pre></div>
<div id="parameter-settings" class="section level2">
<h2>1. Parameter Settings</h2>
<p>Define your variables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pat &lt;-<span class="st"> 'CTANNCAGNNCTTNNCGANNCTANNCTTNNGGANNCTANNCAGNNCTTNNCGANNCTANNCTTNNGGANNCTANNCAGNN'</span> <span class="co"># matching pattern (BC32)</span>
restriction &lt;-<span class="st"> 'CTCGAG'</span> <span class="co"># The restriction sequence flanking the barcode pattern at the 3'end</span>
idx_mis &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># number of mismatch allowed in index</span>
base_q &lt;-<span class="st"> </span><span class="dv">20</span> <span class="co"># minimum average base quality score in first 90 nucleotides</span>
bb_mis &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># number of mismatch allowed in barcode backbone sequence</span>
indels &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># total edit distance resulting from indels that are tolerated for barcode matching (try to keep this low)</span>
dl &lt;-<span class="st"> </span><span class="dv">12</span> <span class="co"># threshold for Damerau-Levenshtein distance in step pooling similar sequences</span>
min.count &lt;-<span class="st"> </span><span class="dv">10</span> <span class="co"># in the final merged table, replace barcodes with occurence below min.count (considered as noise) with 0</span></code></pre></div>
<p>Next, you need to prepare a data frame containing minimal information about the data to be analyzed. In this Vignette an example is provided through a textfile named ‘sampname.txt’ in which each row of the file contains 3 tab-separated columns indicating the sample names, paths of the corresponding sequencing fastq file and the sequences of the corresponding multiplexing index.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampname &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">'sampname.txt'</span>, <span class="co"># sampname provides a list of sample names, matching sequencing files and multiplexing index</span>
                       <span class="dt">header =</span> F,
                       <span class="dt">stringsAsFactor =</span> F)
<span class="kw">colnames</span>(sampname) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'sample'</span>, <span class="st">'file'</span>, <span class="st">'index'</span>)</code></pre></div>
</div>
<div id="generate-summaries-for-each-sample" class="section level2">
<h2>2.Generate summaries for each sample</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bc_data &lt;-<span class="st"> </span><span class="kw">generateSummaries</span>(<span class="dt">pat =</span> pat,
                             <span class="dt">restriction =</span> restriction,
                             <span class="dt">sampname =</span> sampname,
                             <span class="dt">base_q =</span> base_q,
                             <span class="dt">idx_mis =</span> idx_mis,
                             <span class="dt">bb_mis =</span> bb_mis,
                             <span class="dt">indels =</span> indels)

<span class="kw">save</span>(bc_data, <span class="st">'bc_data.Rdata) # Keep this object, it is required for downstream processing!</span></code></pre></div>
<p>This function call will generate (in the working directory) a table of raw counts for each sample, and a series of additional graphs and tables for quality control. The top ten sequences failing the index or barcode matching are listed in the textfiles containing the suffixes “_no_index&quot; or “_no_bc_match&quot; respectively. These files also show the alignment of the sequences with the index (respectively barcode pattern) to easily identify why it didn’t match.</p>
</div>
<div id="add-invited-sequences" class="section level2">
<h2>3. Add invited sequences</h2>
<p>After careful QC review, you can provide a list of sequences that you think deserve to be included in the analysis (identified in the “_no_index&quot; and “_no_bc_match&quot; files). invited_idx contains sequences that failed the index matching but you believe are worthy to be included in the analysis (let’s say if the index did not match because of one substitution), this dataframe contains 2 columns, with the sequence in the first column and the sample name in the second column. invided_bc contains sequences that failed the barcode matching (for example because of a big deletion), in the form of a vector (or in a data frame with one column). Be careful, this step directly modifies the raw table of counts, so make sure you only run it once!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">invited_idx &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">'invited_idx.txt'</span>, <span class="dt">header =</span> F, <span class="dt">stringsAsFactor =</span> F) <span class="co"># format: column 1 -&gt; sequences, column 2 -&gt; sample name</span>
invited_bc &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">'invited_bc.txt'</span>, <span class="dt">header =</span> F, <span class="dt">stringsAsFactor =</span> F) <span class="co"># format: column 1 -&gt; sequences</span>

<span class="kw">addInvitedSeq</span>(<span class="dt">pat =</span> pat,
              <span class="dt">bc_data =</span> bc_data,
              <span class="dt">dir =</span> <span class="kw">getwd</span>(),
              <span class="dt">invited_idx =</span> invited_idx,
              <span class="dt">invited_bc =</span> invited_bc)</code></pre></div>
</div>
<div id="pool-similar-barcodes-together-according-to-damerau-levenshtein-distance" class="section level2">
<h2>4. Pool similar barcodes together (according to Damerau-Levenshtein distance)</h2>
<p>This function will generate a new table of counts, where similar sequences are pooled together under one unique barcode sequence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">poolBC</span>(<span class="dt">sampname =</span> sampname,
        <span class="dt">dir =</span> <span class="kw">getwd</span>(),
        <span class="dt">dl =</span> dl)</code></pre></div>
</div>
<div id="merge-data" class="section level2">
<h2>5. Merge data</h2>
<p>In this last step, all samples are merged together into one table to facilitate the analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts &lt;-<span class="st"> </span><span class="kw">mergeSummaries</span>(<span class="dt">sampname =</span> sampname,
                         <span class="dt">dir =</span> <span class="kw">getwd</span>(),
                         <span class="dt">min.count =</span> min.count)

freq &lt;-<span class="st"> </span><span class="kw">countsToFreq</span>(<span class="dt">counts =</span> counts)

cpm &lt;-<span class="st"> </span><span class="kw">countsToCPM</span>(<span class="dt">counts =</span> counts)

<span class="kw">write.table</span>(counts, <span class="st">'merged_summary_pooled_count.txt'</span>, <span class="dt">sep=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">row.names =</span> F)
<span class="kw">write.table</span>(freq, <span class="st">'merged_summary_pooled_freq.txt'</span>, <span class="dt">sep=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">row.names =</span> F)
<span class="kw">write.table</span>(cpm, <span class="st">'merged_summary_pooled_cpm.txt'</span>, <span class="dt">sep=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, <span class="dt">row.names =</span> F)

<span class="co"># export session info for reproducibility</span>
<span class="kw">writeLines</span>(<span class="kw">capture.output</span>(<span class="kw">sessionInfo</span>()), <span class="st">'sessionInfo.txt'</span>)</code></pre></div>
</div>
<div id="explore-data" class="section level2">
<h2>6. Explore data</h2>
<p>This function starts a shiny app to explore the table of counts and frequencies generated in the previous step.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">exploreBC</span>(<span class="kw">getwd</span>())</code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
